Content.Server/GameObjects/Components/PipeDispenserComponent.cs

```diff
@@ -0,0 +1,196 @@
+        }
+
+        private void UpdatePower(object sender, PowerStateEventArgs args)
+        {
+            var state = args.Powered ? PipeDispenserVisualState.Normal : PipeDispenserVisualState.Off;
+            TrySetVisualState(state);
+        }
+
+        private void UserInterfaceOnOnReceiveMessage(ServerBoundUserInterfaceMessage serverMsg)
+        {
+            if (!Powered)
+                return;
+
+            var message = serverMsg.Message;
+            switch (message)
+            {
+                case PipeDispenserEjectMessage msg:
+                    TryEject(msg.ID, msg.Amount);
+                    break;
+                case InventorySyncRequestMessage _:
+                    UserInterface?.SendMessage(new PipeDispenserInventoryMessage(Inventory));
+                    break;
+            }
+        }
+
+        private void TryEject(string id, uint amount)
+        {
+            if (_ejecting || _broken)
+            {
+                return;
+            }
+
+            var entry = Inventory.Find(x => x.ID == id);
+            if (entry == null)
+            {
+                return;
+            }
+
+            _ejecting = true;
+            TrySetVisualState(PipeDispenserVisualState.Eject);
+
+            Timer.Spawn(_ejectDelay, () =>
+            {
+                _ejecting = false;
+                TrySetVisualState(PipeDispenserVisualState.Normal);
+
+                var position = Owner.Transform.Coordinates;
+                if (_accessDirection != Direction.Invalid)
+                    position = position.Offset(_accessDirection.ToVec());
+
+                for (var i = 0; i < amount; i++)
+                {
+                    var e = Owner.EntityManager.SpawnEntity(id, position);
+                    if(e.TryGetComponent(out CollidableComponent collidable))
+                        collidable.Anchored = false;
+                }
+            });
+
+            EntitySystem.Get<AudioSystem>().PlayFromEntity(_soundVend, Owner, AudioParams.Default.WithVolume(-2f));
+        }
+
+        private void TrySetVisualState(PipeDispenserVisualState state)
+        {
+            var finalState = state;
+            if (_broken)
+            {
+                finalState = PipeDispenserVisualState.Broken;
+            }
+            else if (_ejecting)
+            {
+                finalState = PipeDispenserVisualState.Eject;
+            }
+            else if (!Powered)
+            {
+                finalState = PipeDispenserVisualState.Off;
+            }
+
+            if (Owner.TryGetComponent(out AppearanceComponent appearance))
+            {
+                appearance.SetData(PipeDispenserVisuals.VisualState, finalState);
+            }
+        }
+
+        private Direction DirectionTo(IEntity other)
+        {
+            return (other.Transform.WorldPosition - Owner.Transform.WorldPosition).GetDir();
+        }
+
+        public void OnBreak(BreakageEventArgs eventArgs)
```

Comment:
I think you should disable all buttons on the client/close the menu if the dispenser breaks/has no power